### ISSUE TITLE
Auto-Generate Transformation Field Mappings with AI

### ISSUE THREAD

**Issue by JacobJNilsson** [JacobJNilsson/data-contract-generator#5]
# Auto-Generate Transformation Field Mappings with AI

## Problem

Creating transformation contracts is currently **manual and tedious**:

1. User generates source contract (e.g., CSV with 50 columns)
2. User generates destination contract (e.g., database with 30 columns)
3. User must **manually map** each destination field to source field(s)
4. User must **manually write** transformation logic (CONCAT, CAST, etc.)
5. User must **manually define** validation and enrichment rules

This is:
- **Time-consuming** - Can take hours for complex schemas
- **Error-prone** - Easy to miss fields or make typos
- **Repetitive** - Many mappings are obvious (name → name)
- **Not scalable** - Worse with 100+ column tables

**This is exactly where AI shines** - pattern matching, schema analysis, and code generation.

### 1. Direct Field Mapping

**Source:** `first_name`, `last_name`, `email_address`
**Destination:** `name`, `email`

**AI suggests:**
```json
{
  "field_mappings": {
    "name": "CONCAT(first_name, ' ', last_name)",
    "email": "email_address"
  }
}
```

### 2. Type Conversions

**Source:** `created_date` (text: "2024-01-15")
**Destination:** `created_at` (datetime)

**AI suggests:**
```json
{
  "field_mappings": {
    "created_at": "CAST(created_date AS TIMESTAMP)"
  }
}
```

### 3. Derived Fields

**Source:** `price`, `quantity`
**Destination:** `total_amount`

**AI suggests:**
```json
{
  "field_mappings": {
    "total_amount": "price * quantity"
  }
}
```

### 4. Normalization

**Source:** `country_name` (text: "United States")
**Destination:** `country_code` (text: "US")

**AI suggests:**
```json
{
  "field_mappings": {
    "country_code": "LOOKUP(country_name, country_mapping_table)"
  },
  "enrichment": {
    "lookup_tables": ["country_mapping_table"]
  }
}
```

### 5. Complex Transformations

**Source:** `full_address` (text: "123 Main St, New York, NY 10001")
**Destination:** `street`, `city`, `state`, `zip`

**AI suggests:**
```json
{
  "field_mappings": {
    "street": "SPLIT_PART(full_address, ',', 1)",
    "city": "TRIM(SPLIT_PART(full_address, ',', 2))",
    "state": "SPLIT_PART(SPLIT_PART(full_address, ',', 3), ' ', 1)",
    "zip": "SPLIT_PART(SPLIT_PART(full_address, ',', 3), ' ', 2)"
  }
}
```

## Proposed Solution

Add AI-powered transformation generation:

```python
generate_transformation_with_ai(
    source_contract: SourceContract | str,  # Contract or path
    destination_contract: DestinationContract | str,
    transformation_id: str,
    llm_provider: str = "openai",  # "openai", "anthropic", "local"
    model: str | None = None,  # Model name (default: gpt-4, claude-3.5-sonnet)
    confidence_threshold: float = 0.8,  # Only include high-confidence mappings
    include_explanations: bool = True,  # Add comments explaining mappings
    config: dict[str, Any] | None = None
) -> TransformationContract
```

### Example Usage

```python
# Basic usage
transformation = generate_transformation_with_ai(
    source_contract="contracts/source_sales.json",
    destination_contract="contracts/dest_warehouse.json",
    transformation_id="sales_to_warehouse"
)

# Returns transformation with suggested mappings:
{
  "transformation_id": "sales_to_warehouse",
  "source_ref": "source_sales",
  "destination_ref": "dest_warehouse",
  "field_mappings": {
    "order_id": "id",
    "customer_name": "CONCAT(first_name, ' ', last_name)",
    "order_total": "price * quantity",
    "order_date": "CAST(date AS TIMESTAMP)"
  },
  "transformations": {
    "order_date": {
      "type": "cast",
      "from_type": "text",
      "to_type": "timestamp",
      "format": "YYYY-MM-DD"
    }
  },
  "metadata": {
    "ai_generated": true,
    "confidence_scores": {
      "order_id": 0.95,
      "customer_name": 0.90,
      "order_total": 0.85,
      "order_date": 0.88
    },
    "explanations": {
      "customer_name": "Combined first_name and last_name fields",
      "order_total": "Calculated from price and quantity"
    }
  }
}
```

### CLI Integration

```bash
# Auto-generate transformation
contract-gen transform \
  --source contracts/source.json \
  --destination contracts/dest.json \
  --auto-map \
  --model gpt-4 \
  --output contracts/transform.json

# With confidence threshold
contract-gen transform \
  --source contracts/source.json \
  --destination contracts/dest.json \
  --auto-map \
  --confidence 0.9 \  # Only include very confident mappings
  --review  # Open in editor for review before saving
```

### Approach 1: LLM-Based Mapping (Recommended)

Use GPT-4, Claude, or local LLMs to analyze contracts and suggest mappings.

#### Prompt Engineering

```python
SYSTEM_PROMPT = """
You are an expert data engineer analyzing source and destination schemas
to create field mappings for data transformation.

Your task:
1. Match destination fields to source field(s)
2. Suggest transformation logic (CONCAT, CAST, calculations)
3. Identify fields that need enrichment (lookups, defaults)
4. Provide confidence scores (0-1) for each mapping

Rules:
- Direct matches (same name): confidence 1.0
- Similar names (first_name -> name): confidence 0.9
- Derived fields (price * quantity): confidence 0.8
- Complex logic: confidence 0.6-0.7
- Return JSON only, no explanations outside JSON
"""

USER_PROMPT = """
Source Schema:
{source_fields_with_types}

Destination Schema:
{destination_fields_with_types}

Sample Source Data:
{sample_data}

Generate field mappings in this format:
{{
  "field_mappings": {{
    "dest_field": "source_expression"
  }},
  "confidence_scores": {{
    "dest_field": 0.95
  }},
  "explanations": {{
    "dest_field": "Why this mapping"
  }}
}}
"""
```

#### Implementation

```python
from openai import OpenAI

def generate_ai_mappings(
    source_contract: SourceContract,
    destination_contract: DestinationContract,
    llm_provider: str = "openai"
) -> dict:
    # Prepare prompt data
    source_info = format_schema_for_llm(source_contract)
    dest_info = format_schema_for_llm(destination_contract)
    samples = source_contract.quality_metrics.sample_data[:5]

    # Call LLM
    client = OpenAI()
    response = client.chat.completions.create(
        model="gpt-4",
        messages=[
            {"role": "system", "content": SYSTEM_PROMPT},
            {"role": "user", "content": USER_PROMPT.format(
                source_fields_with_types=source_info,
                destination_fields_with_types=dest_info,
                sample_data=samples
            )}
        ],
        response_format={"type": "json_object"}
    )

    # Parse response
    mappings = json.loads(response.choices[0].message.content)
    return mappings
```

### Approach 2: Rule-Based + ML (Alternative)

Combine heuristics with trained models:

1. **Direct match**: Exact field name match
2. **Fuzzy match**: Similar names (Levenshtein distance)
3. **Type-aware**: Match fields of compatible types
4. **Pattern detection**: Learn from historical mappings
5. **Semantic similarity**: Use embeddings (field names, descriptions)

```python
def rule_based_mapping(source, destination):
    mappings = {}

    for dest_field in destination.fields:
        # Try exact match
        if dest_field in source.fields:
            mappings[dest_field] = dest_field
            continue

        # Try fuzzy match
        best_match = find_fuzzy_match(dest_field, source.fields)
        if best_match and similarity > 0.8:
            mappings[dest_field] = best_match
            continue

        # Try semantic match
        semantic_match = find_semantic_match(dest_field, source.fields)
        if semantic_match:
            mappings[dest_field] = semantic_match
```

**Recommendation**: Start with LLM-based approach (simpler, better results), add rule-based fallback for offline mode.

### Confidence Scores

Assign confidence to each mapping:

- **1.0** - Exact field name match
- **0.9-0.95** - Fuzzy name match (very similar)
- **0.8-0.9** - Type-aware match + semantic similarity
- **0.7-0.8** - Derived fields with clear logic
- **0.5-0.7** - Complex transformations
- **< 0.5** - Uncertain, needs human review

Only include mappings above `confidence_threshold` (default: 0.8).

### Human Review Workflow

```python
# Generate with AI
transformation = generate_transformation_with_ai(...)

# Review uncertain mappings
uncertain = [
    (field, mapping, score)
    for field, mapping in transformation.field_mappings.items()
    if transformation.metadata["confidence_scores"][field] < 0.8
]

if uncertain:
    print("⚠️  Please review these mappings:")
    for field, mapping, score in uncertain:
        print(f"  {field} <- {mapping} (confidence: {score:.2f})")
        # Prompt for confirmation/edit
```

### Interactive Mode

```bash
# Interactive review
contract-gen transform \
  --source contracts/source.json \
  --destination contracts/dest.json \
  --auto-map \
  --interactive

# Output:
✓ order_id <- id (confidence: 0.95)
✓ customer_name <- CONCAT(first_name, ' ', last_name) (confidence: 0.90)
? product_code <- sku (confidence: 0.75)
  Accept? [y/n/edit]: y
✓ Accepted
```

### Phase 1: Basic AI Mapping (MVP)

- [ ] LLM integration (OpenAI GPT-4)
- [ ] Prompt engineering for field mapping
- [ ] Generate simple field mappings (1:1)
- [ ] Confidence scores
- [ ] CLI `--auto-map` flag

### Phase 2: Advanced Transformations

- [ ] Type conversion suggestions
- [ ] Derived field detection (calculations)
- [ ] CONCAT suggestions for name fields
- [ ] Handle nested objects (flatten/join)
- [ ] Multiple LLM providers (Anthropic, local)

### Phase 3: Human-in-the-Loop

- [ ] Interactive review mode
- [ ] Confidence threshold filtering
- [ ] Explanation generation
- [ ] Learning from human corrections

### Phase 4: Advanced Features

- [ ] Rule-based fallback (offline mode)
- [ ] Historical mapping learning
- [ ] Domain-specific patterns (e.g., address parsing)
- [ ] Validation rule suggestions
- [ ] Enrichment strategy suggestions

## Dependencies

```toml
[project.optional-dependencies]
ai = [
    "openai>=1.0.0",
    "anthropic>=0.18.0",
    "instructor>=1.0.0",  # Structured outputs
]
```

### LLM API Costs

- **GPT-4**: ~$0.01-0.03 per transformation
- **GPT-3.5**: ~$0.001-0.002 per transformation
- **Claude 3.5 Sonnet**: ~$0.01-0.02 per transformation

For typical schemas (20-50 fields), cost is negligible. For 100+ field schemas, may need optimization.

### Cost Optimization

1. **Cache LLM responses** - Same schema pairs = same mappings
2. **Use cheaper models** - GPT-3.5 for simple cases, GPT-4 for complex
3. **Batch processing** - Generate multiple transformations in one call
4. **Local models** - Use Llama 3, Mistral for offline/privacy needs

## Acceptance Criteria

**Core:**
- [ ] Can generate field mappings using OpenAI GPT-4
- [ ] Handles simple 1:1 field mappings
- [ ] Suggests CONCAT for name fields
- [ ] Suggests type conversions
- [ ] Returns confidence scores
- [ ] Only includes high-confidence mappings (> threshold)

**Advanced:**
- [ ] Detects derived fields (calculations)
- [ ] Suggests complex transformations (string parsing)
- [ ] Provides explanations for each mapping
- [ ] Interactive review mode
- [ ] Multiple LLM provider support

**CLI:**
- [ ] `--auto-map` flag works
- [ ] `--model` flag to choose LLM
- [ ] `--confidence` threshold setting
- [ ] `--interactive` review mode

### Data Handling

⚠️ **Important**: Source data samples are sent to LLM APIs

**Options:**
1. **Anonymize samples** - Replace values with placeholders
2. **Schema-only mode** - Don't send sample data
3. **Local LLM** - Use on-prem model (Llama, Mistral)
4. **User consent** - Prompt before sending data to API

**Implementation:**
```python
# Default: anonymize samples
generate_transformation_with_ai(
    ...,
    anonymize_samples=True  # Default
)

# Schema-only (no samples sent)
generate_transformation_with_ai(
    ...,
    send_samples=False
)

# Local LLM (no external API)
generate_transformation_with_ai(
    ...,
    llm_provider="local",
    model="llama-3-70b"
)
```

### 1. Fine-Tuned Model

Train a specialized model on transformation datasets:
- **Pros**: Better accuracy, lower cost per inference
- **Cons**: Need training data, maintenance overhead

### 2. Semantic Embeddings

Use embedding models (Sentence Transformers) for field matching:
- **Pros**: Fast, cheap, no API calls
- **Cons**: Limited to field name matching, can't generate complex logic

### 3. Knowledge Graph

Build a knowledge graph of common field relationships:
- **Pros**: Deterministic, explainable
- **Cons**: Requires curation, doesn't generalize

**Recommendation**: Start with LLM (Phase 1), add embeddings as fallback (Phase 2).

## Nice to Have

- Visual mapping editor (web UI)
- Learn from user corrections over time
- Domain-specific mapping templates (e-commerce, healthcare, etc.)
- Suggest data quality rules based on analysis
- Generate SQL/Python code for transformations
- Integration with dbt, Airflow DAGs
- A/B testing different mapping strategies

## Related Issues

- #4 - CLI Tool for Contract Generation

## Labels

`enhancement`, `ai`, `transformation`, `high-priority`, `automation`

**Comment by JacobJNilsson** [comment]
We do not want the CLI tool to be responsible for the transformation and mapping decisions. This should be done by the user, and the tool can generate the contract skeleton, and verify correctness.
